timings/lib/src/timings.dart

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:timings/src/clock.dart';
+
+/// The timings of an operation, including its [startTime], [stopTime], and
+/// [duration].
+abstract class TimeSlice {
+  /// The total duration of this operation, equivalent to taking the difference
+  /// between [stopTime] and [startTime].
+  Duration get duration => stopTime?.difference(startTime);
+
+  DateTime get startTime;
+
+  DateTime get stopTime;
+
+  TimeSlice._();
+
+  factory TimeSlice(DateTime startTime, DateTime stopTime) =>
+      _TimeSliceImpl(startTime, stopTime);
+
+  @override
+  String toString() => '($startTime + $duration)';
+}
+
+class _TimeSliceImpl extends TimeSlice {
+  @override
+  final DateTime startTime;
+
+  @override
+  final DateTime stopTime;
+
+  _TimeSliceImpl(this.startTime, this.stopTime) : super._();
+}
+
+/// The timings of an async operation, consist of several sync [slices] and
+/// includes total [startTime], [stopTime], and [duration].
+abstract class TimeSliceGroup extends TimeSlice {
+  List<TimeSlice> get slices;
+
+  @override
+  DateTime get startTime => slices.first?.startTime;
+
+  @override
+  DateTime get stopTime => slices.last?.stopTime;
+
+  /// Sum of [duration]s of all [slices].
+  ///
+  /// If some of slices implements [TimeSliceGroup] [innerDuration] will be used
+  /// to compute sum.
+  Duration get innerDuration => slices.fold(
+      Duration.zero,
+      (duration, slice) =>
+          duration +
+          (slice is TimeSliceGroup ? slice.innerDuration : slice.duration));
+
+  TimeSliceGroup._() : super._();
+
+  factory TimeSliceGroup() => _TimeSliceGroupImpl();
+
+  @override
+  String toString() => slices.toString();
+}
+
+class _TimeSliceGroupImpl extends TimeSliceGroup {
+  @override
+  final slices = [];
+
+  _TimeSliceGroupImpl() : super._();
+}
+
+abstract class TimeTracker implements TimeSlice {
```

Comment:
It's strange to me that a Tracker would be a Slice...

Overall this library has a lot more `extends` and `implements` than I would hope for. Why is a group of slices also a single slice? Why do we need all the `*Impl` classes?